src/dispatch/experimental/durable/frame.c

```diff
@@ -239,12 +268,20 @@ static PyObject *set_frame_ip(PyObject *self, PyObject *args) {
 }
 
 static PyObject *set_frame_sp(PyObject *self, PyObject *args) {
-    PyObject *frame_obj;
+    PyObject *obj;
     int sp;
-    if (!PyArg_ParseTuple(args, "Oi", &frame_obj, &sp)) {
+    if (!PyArg_ParseTuple(args, "Oi", &obj, &sp)) {
         return NULL;
     }
-    InterpreterFrame *frame = get_interpreter_frame(frame_obj);
+    PyGenObject *gen_like = get_generator_like_object(obj);
+    if (!gen_like) {
+        return NULL;
+    }
+    if (gen_like->gi_frame_state >= FRAME_CLEARED) {
+        PyErr_SetString(PyExc_RuntimeError, "Cannot mutate cleared frame");
+        return NULL;
+    }
```

Comment:
Is it worth moving this check into a macro so we don't repeat it over and over? (fine if you like it explicit better, it just seems to be a common pattern in CPython)

Reply:
I think we should avoid a macro in this case. Macros that contain non-local control flow (`return NULL` in this case) are confusing, and it's not a lot of duplicated code that it would eliminate.